rfcs/2021-11-12-9811-vrl-vm.md

```diff
@@ -0,0 +1,347 @@
+function `call`.
+
+The `ArgumentList` parameter that is passed into `call` will have access to the
+parameter stack and the parameter list exposed by the `Function`. This can use
+these to return the appropriate values for the `required`, `optional` etc..
+functions.
+
+Since the `compile` function was also used by some `stdlib` functions to
+validate the parameters we may alse need a function `validate_parameters` that
+these functions can use to perform extra validation on the parameters.
+
+### Debugging
+
+We need debugging tools to help us introspect and debug the compiled bytecode.
+
+#### Dissassembler
+
+Output the compiled code in readable format to the console.
+
+#### Step debugger
+
+Output each step of the VM to the console, print the Stack and other state
+variables in the VM. Allow the user to press a key to move to the next step.
+
+#### Linter
+
+Check each instruction to ensure:
+
+- each Jump Opcode jumps to a valid location in the bytecode.
+- each constant references a valid index in the constants list
+- each path index references a valid path.
+
+It should be possible to run the linter after every compile (at Vector boot
+time), so the linter should run relatively fast.
+
+### Testing
+
+The VM will need thorough testing since it is possible to generate invalid
+bytecode. For example, the following bytecode is invalid:
+
+```rust
+OpCode(CONSTANT)
+Primitive(1)
+OpCode(ADD)
+```
+
+The `ADD` OpCode expects there to be two values on the stack to add
+together. We should never get into the situation where the VRL compiler can
+generate this.
+
+With property testing we can generate a vast combination of valid VRL ASTs,
+compile this to bytecode and then run the VM to ensure we get no buffer
+underflows, invalid constant locations or other invalid bytecode.
+
+Given that implementing the VM involves adding methods to the Expression trait,
+it should be possible to run the existing AST alongside the VM. We can then
+ensure that we get exactly the same results from each approach.
+
+## Rationale
+
+With each node of the AST compiled down to just a few bytes and all
+instructions held in contiguous memory evaluation of the program should be able
+to take full advantage of the CPU cache which should result in much faster
+execution.
+
+## Drawbacks
+
+Downsides to using a Vm:
+
+- The code is a bit more complex. With an AST that is walked it is fairly
+  apparent what the code will be doing at any point. With a Vm, this is not the
+  case, it is harder to look at the instructions in the Vm and follow back to
+  what part of the VRL code is being evaluated. We will need to write some
+  extensive debugging tools to allow for decent introspection into the Vm.
+
+  However, VRL is a very simple language, which does also mean the VM will be
+  simple.
+
+- We lose some safety that we get from the Rust compiler. There will need
+  to be significant fuzz testing to ensure that the code runs correctly under
+  all circumstances.
```

Comment:
Core team will help you get the CI infra in place for this, spec out the fuzz tests.